File: src/potential/inverse_power_coulomb_bounding_potential.py
# JeLLFysh - a Python application for all-atom event-chain Monte Carlo - https://github.com/jellyfysh
# Copyright (C) 2019 The JeLLyFysh organization
# (see the AUTHORS file for the full list of authors)
#
# This file is part of JeLLyFysh.
#
# JeLLyFysh is free software: you can redistribute it and/or modify it under the terms of the GNU General Public
# License as published by the Free Software Foundation, either > version 3 of the License, or (at your option) any
# later version.
#
# JeLLyFysh is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
# warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with JeLLyFysh in the LICENSE file.
# If not, see <https://www.gnu.org/licenses/>.
#
# If you use JeLLyFysh in published work, please cite the following reference (see [Hoellmer2019] in References.bib):
# Philipp Hoellmer, Liang Qin, Michael F. Faulkner, A. C. Maggs, Werner Krauth
# JeLLyFysh-Version1.0 -- a Python application for all-atom event-chain Monte Carlo,
# arXiv e-prints: 1907.12502 (2019), https://arxiv.org/abs/1907.12502
#
"""Module for the InversePowerCoulombBoundingPotential."""
import logging
from typing import MutableSequence, Sequence
from base.exceptions import ConfigurationError
from base.logging import log_init_arguments
from base import vectors
from setting import hypercubic_setting as setting
from .potential import InvertiblePotential
from .inverse_power_potential import InversePowerPotential


# noinspection PyMethodOverriding
class InversePowerCoulombBoundingPotential(InvertiblePotential):
    """
    This class implements the pair potential U_ij = c_i * c_j * k / |r_ij,0|.

    k is a prefactor, c_i and c_j are the charges of the involved units and r_ij,0 = nearest(r_j - r_i) is the
    shortest separation between the units, possible corrected for periodic boundaries. With a correct prefactor, this
    potential can bound the merged image Coulomb potential. This class assumes that i is the active unit.
    Any constant k > 1.5836 is appropriate for a bounding potential in a hypercubic box. This class is only
    implemented for a hypercubic setting.
    """

    def __init__(self, prefactor: float = 1.5837) -> None:
        """
        The constructor of the InversePowerCoulombBoundingPotential class.

        Parameters
        ----------
        prefactor : float, optional
            The prefactor k of the potential.

        Raises
        ------
        base.exceptions.ConfigurationError
            If the hypercubic setting is not initialized.
        """
        log_init_arguments(logging.getLogger(__name__).debug, self.__class__.__name__,
                           prefactor=prefactor)
        if not setting.initialized():
            raise ConfigurationError("Potential {0} can only be used in a hypercubic setting."
                                     .format(self.__class__.__name__))
        super().__init__(prefactor=prefactor)
        self._inverse_power_potential = InversePowerPotential(1, prefactor)

    def _potential(self, charge_product: float, separation: Sequence[float]) -> float:
        """Return the potential for the given separation and charge product."""
        return self._inverse_power_potential.potential(charge_product, separation)

    def derivative(self, direction: int, separation: Sequence[float], charge_one: float, charge_two: float) -> float:
        """
        Return the derivative of the potential along a direction.

        Parameters
        ----------
        direction : int
            The direction of the derivative.
        separation : Sequence[float]
            The separation vector r_ij.
        charge_one : float
            The charge c_i.
        charge_two : float
            The charge c_j.

        Returns
        -------
        float
            The derivative.
        """
        return self._inverse_power_potential.derivative(direction, separation, charge_one, charge_two)

    def displacement(self, direction: int, separation: MutableSequence[float], charge_one: float, charge_two: float,
                     potential_change: float) -> float:
        """
        Return the displacement of the active unit i until the cumulative event rate of the potential equals the given
        potential change.

        This method computes the cumulative event rate when displacing the active unit by the system length. For any
        leftover of the potential change, the displacement is then computed separately by using an instance of an
        inverse power potential. The leftover displacement must then be smaller than a system length.

        Parameters
        ----------
        direction : int
            The direction of motion of the active unit i.
        separation : MutableSequence[float]
            The separation vector r_ij.
        charge_one : float
            The charge c_i.
        charge_two : float
            The charge c_j.
        potential_change : float
            The sampled potential change.

        Returns
        -------
        float
            The displacement of the active unit i where the cumulative event rate equals the sampled potential change.

        Raises
        ------
        AssertionError
            If the leftover displacement is larger than or equal to a system length.
        """
        charge_product = charge_one * charge_two
        prefactor_product = self._prefactor * charge_product

        potential_zero = self._potential(charge_product,
                                         vectors.copy_vector_with_replaced_component(separation, direction, 0.0))
        separation_half_length = vectors.copy_vector_with_replaced_component(separation, direction,
                                                                             setting.system_length_over_two)
        potential_half_length = self._potential(charge_product, separation_half_length)

        potential_change_per_system_length = abs(potential_zero - potential_half_length)
        multiple_displacement, potential_change = divmod(potential_change, potential_change_per_system_length)
        multiple_displacement *= setting.system_length

        remaining_displacement = self._inverse_power_potential.displacement(direction, list(separation),
                                                                            charge_one, charge_two, potential_change)

        # Displacing active leaf unit by remaining_displacement would carry it into interaction with next periodic image
        # Place active leaf unit so that separation in direction of motion is L/2 and calculate again
        if separation[direction] - remaining_displacement < -setting.system_length_over_two:
            current_potential = self._inverse_power_potential.potential(charge_product, separation)
            if separation[direction] <= 0:
                potential_change -= max(0.0, potential_half_length - current_potential)
            elif prefactor_product > 0:
                potential_change -= potential_zero - current_potential
            else:
                raise RuntimeError('Remaining energy change greater than potential change per system length.')
            remaining_displacement = setting.system_length_over_two + separation[direction]
            separation[direction] = setting.system_length_over_two
            remaining_displacement += self._inverse_power_potential.displacement(direction, list(separation),
                                                                                 charge_one, charge_two,
                                                                                 potential_change)
        assert remaining_displacement < setting.system_length

        return multiple_displacement + remaining_displacement

File: jellyfysh/potential/inverse_power_coulomb_bounding_potential/inverse_power_coulomb_bounding_potential.c
/************************************************************************************************************************
 * JeLLFysh - a Python application for all-atom event-chain Monte Carlo - https://github.com/jellyfysh                  *
 * Copyright (C) 2019, 2022 The JeLLyFysh organization                                                                  *
 * (See the AUTHORS.md file for the full list of authors.)                                                              *
 *                                                                                                                      *
 * This file is part of JeLLyFysh.                                                                                      *
 *                                                                                                                      *
 * JeLLyFysh is free software: you can redistribute it and/or modify it under the terms of the GNU General Public       *
 * License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later *
 * version.                                                                                                             *
 *                                                                                                                      *
 * JeLLyFysh is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied      *
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more        *
 * details.                                                                                                             *
 *                                                                                                                      *
 * You should have received a copy of the GNU General Public License along with JeLLyFysh in the LICENSE file.          *
 * If not, see <https://www.gnu.org/licenses/>.                                                                         *
 *                                                                                                                      *
 * If you use JeLLyFysh in published work, please cite the following reference (see [Hoellmer2020] in References.bib):  *
 * Philipp Hoellmer, Liang Qin, Michael F. Faulkner, A. C. Maggs, and Werner Krauth,                                    *
 * JeLLyFysh-Version1.0 -- a Python application for all-atom event-chain Monte Carlo,                                   *
 * Computer Physics Communications, Volume 253, 107168 (2020), https://doi.org/10.1016/j.cpc.2020.107168.               *
 ************************************************************************************************************************/

/** @file inverse_power_coulomb_bounding_potential.c
 *  @brief Definitions of functions to compute the directional time derivative of the inverse power coulomb bounding
 *         potential along a given velocity vector of the active unit, and to compute the required time displacement
 *         along its velocity where the cumulative event rate equals a sampled potential change.
 *
 *  The inverse power coulomb bounding potential between a target unit j and an active unit i is given by
 *  U_ij = c_i * c_j * k / |r_ij,0|. Here, r_ij,0 = nearest(r_j - r_i) is the minimum separation vector, that is, the
 *  vector between r_i and the closest image of r_j under periodic boundary conditions. The charges of the units are
 *  c_i and c_j, respectively, and k is a prefactor. The functions in this file are explicitly implemented for a cubic
 *  setting with side length L in three dimensions.
 *
 *  @author The JeLLyFysh organization.
 *  @bug No known bugs.
 */
#include "inverse_power_coulomb_bounding_potential.h" // Include declarations.

#include <math.h> // For fabs, INFINITY, pow, sqrt.
#include <stdbool.h> // For true, false.


/** @brief Return the directional time derivative along a given velocity vector of the active unit for the given
 *         separation r_ij,0.
 *
 *  @param prefactor_product The product c_i * c_j * k.
 *  @param velocity The velocity of the active unit.
 *  @param separation The current separation r_ij,0.
 *  @return The directional time derivative.
 */
double derivative(double prefactor_product, double velocity[3], double separation[3]) {
    double constant_factor = prefactor_product / pow(separation[0] * separation[0] + separation[1] * separation[1]
                                                     + separation[2] * separation[2], 3.0 / 2.0);
    return (velocity[0] * separation[0] + velocity[1] * separation[1] + velocity[2] * separation[2]) * constant_factor;
}


/** @brief  Return the gradient of the inverse power coulomb bounding potential evaluated at the given separation.
 *
 *  @param prefactor_product The product c_i * c_j * k.
 *  @param separation The separation r_ij,0.
 *  @return The gradient with respect to the position r_i of the active unit.
 */
struct Gradient gradient(double prefactor_product, double separation[3]) {
    double constant_factor = prefactor_product / pow(separation[0] * separation[0] + separation[1] * separation[1]
                                                     + separation[2] * separation[2], 3.0 / 2.0);
    return (struct Gradient) {constant_factor * separation[0], constant_factor * separation[1],
                              constant_factor * separation[2]};
}


/** @brief Compute the inverse power coulomb bounding potential evaluated at the given separation r_ij,0.
 *
 *  @param prefactor_product The product c_i * c_j * k.
 *  @param separation The current separation r_ij,0.
 *  @return The potential.
 */
double potential(double prefactor_product, double separation[3]) {
    return prefactor_product /
           sqrt(separation[0] * separation[0] + separation[1] * separation[1] + separation[2] * separation[2]);
}


/** @brief Compute the inverse power coulomb bounding potential evaluated at the given separation norm |r_ij,0|.
 *
 *  @param prefactor_product The product c_i * c_j * k.
 *  @param separation_norm The current separation norm |r_ij,0|.
 *  @return The potential.
 */
double potential_from_norm(double prefactor_product, double separation_norm) {
    return prefactor_product / separation_norm;
}


/** @brief Return the required time displacement of the active unit along its velocity where the cumulative event rate
 *         of the potential equals the given potential change.
 *
 *  @param prefactor_product The product c_i * c_j * k.
 *  @param velocity The velocity of the active unit.
 *  @param separation The current separation r_ij,0.
 *  @param potential_change The sampled potential change.
 *  @param system_length The system length L of the cubic setting.
 *  @return The required time displacement.
 */
double displacement(double prefactor_product, double velocity[3], double separation[3], double potential_change,
                    double system_length) {
    double system_length_over_two = system_length / 2.0;
    double velocity_squared = velocity[0] * velocity[0] + velocity[1] * velocity[1] + velocity[2] * velocity[2];
    double separation_squared;
    double separation_dot_velocity;
    bool separation_dot_velocity_positive;
    double current_potential;
    double sqrt_term;
    double displacement_until_next_image;
    double min_displacement_until_next_image;
    int index_next_image;
    double max_displacement;
    double max_potential;
    double new_norm;
    double total_displacement = 0.0;

    if (prefactor_product > 0.0) {
        // Repulsive interaction.
        double min_separation[3];
        // These two variables have to be updated in the loop anytime the separation is changed.
        // This allows to set separation_dot_velocity_positive to true or false without an actual comparison which
        // avoids floating point issues.
        separation_dot_velocity = separation[0] * velocity[0] + separation[1] * velocity[1]
                                  + separation[2] * velocity[2];
        separation_dot_velocity_positive = separation_dot_velocity > 0.0 ? true : false;
        while (1) {
            min_displacement_until_next_image = velocity[0] != 0.0 ?
                                                separation[0] / velocity[0] + system_length_over_two / fabs(velocity[0])
                                                : INFINITY;
            index_next_image = 0;
            displacement_until_next_image = velocity[1] != 0.0 ?
                                            separation[1] / velocity[1] + system_length_over_two / fabs(velocity[1])
                                            : INFINITY;
            if (displacement_until_next_image < min_displacement_until_next_image) {
                min_displacement_until_next_image = displacement_until_next_image;
                index_next_image = 1;
            }
            displacement_until_next_image = velocity[2] != 0.0 ?
                                            separation[2] / velocity[2] + system_length_over_two / fabs(velocity[2])
                                            : INFINITY;
            if (displacement_until_next_image < min_displacement_until_next_image) {
                min_displacement_until_next_image = displacement_until_next_image;
                index_next_image = 2;
            }

            if (separation_dot_velocity_positive) {
                // Norm of separation becomes smaller.
                separation_squared = separation[0] * separation[0] + separation[1] * separation[1]
                                     + separation[2] * separation[2];
                max_displacement = separation_dot_velocity / velocity_squared;
                if (max_displacement <= min_displacement_until_next_image) {
                    min_separation[0] = separation[0] - max_displacement * velocity[0];
                    min_separation[1] = separation[1] - max_displacement * velocity[1];
                    min_separation[2] = separation[2] - max_displacement * velocity[2];
                    current_potential = potential_from_norm(prefactor_product, sqrt(separation_squared));
                    max_potential = potential(prefactor_product, min_separation);
                    if (potential_change >= max_potential - current_potential) {
                        // Active unit can climb potential hill with remaining potential change.
                        // Reduce potential_change and set separation to minimum separation.
                        potential_change -= (max_potential - current_potential);
                        total_displacement += max_displacement;
                        separation[0] = min_separation[0];
                        separation[1] = min_separation[1];
                        separation[2] = min_separation[2];
                        separation_dot_velocity = separation[0] * velocity[0] + separation[1] * velocity[1]
                                                  + separation[2] * velocity[2];
                        // Force that separation_dot_velocity is now negative to avoid floating point issues.
                        separation_dot_velocity_positive = false;
                    } else {
                        new_norm = prefactor_product / (current_potential + potential_change);
                        sqrt_term = separation_dot_velocity * separation_dot_velocity
                                    - velocity_squared * (separation_squared - new_norm * new_norm);
                        total_displacement += (separation_dot_velocity - sqrt(sqrt_term)) / velocity_squared;
                        return total_displacement;
                    }
                } else {
                    min_separation[0] = separation[0] - min_displacement_until_next_image * velocity[0];
                    min_separation[1] = separation[1] - min_displacement_until_next_image * velocity[1];
                    min_separation[2] = separation[2] - min_displacement_until_next_image * velocity[2];
                    current_potential = potential_from_norm(prefactor_product, sqrt(separation_squared));
                    max_potential = potential(prefactor_product, min_separation);
                    if (potential_change >= max_potential - current_potential) {
                        // Active unit can climb potential hill with remaining potential change.
                        // Reduce potential_change and set separation to minimum separation.
                        potential_change -= (max_potential - current_potential);
                        total_displacement += min_displacement_until_next_image;
                        separation[0] = min_separation[0];
                        separation[1] = min_separation[1];
                        separation[2] = min_separation[2];
                        separation[index_next_image] *= -1.0;
                        separation_dot_velocity = separation[0] * velocity[0] + separation[1] * velocity[1]
                                                  + separation[2] * velocity[2];
                        separation_dot_velocity_positive = separation_dot_velocity > 0.0 ? true : false;
                    } else {
                        new_norm = prefactor_product / (current_potential + potential_change);
                        sqrt_term = separation_dot_velocity * separation_dot_velocity
                                    - velocity_squared * (separation_squared - new_norm * new_norm);
                        total_displacement += (separation_dot_velocity - sqrt(sqrt_term)) / velocity_squared;
                        return total_displacement;
                    }
                }
            } else {
                // Norm of separation becomes bigger.
                // Travel down potential hill until interaction with next image.
                total_displacement += min_displacement_until_next_image;
                separation[0] -= min_displacement_until_next_image * velocity[0];
                separation[1] -= min_displacement_until_next_image * velocity[1];
                separation[2] -= min_displacement_until_next_image * velocity[2];
                separation[index_next_image] *= -1.0;
                separation_dot_velocity = separation[0] * velocity[0] + separation[1] * velocity[1]
                                          + separation[2] * velocity[2];
                separation_dot_velocity_positive = separation_dot_velocity > 0.0 ? true : false;
            }
        }
    } else {
        // Attractive interaction.
        double max_separation[3];
        // These two variables have to be updated in the loop anytime the separation is changed.
        // This allows to set separation_dot_velocity_positive to true or false without an actual comparison which
        // avoids floating point issues.
        separation_dot_velocity = separation[0] * velocity[0] + separation[1] * velocity[1]
                                  + separation[2] * velocity[2];
        separation_dot_velocity_positive = separation_dot_velocity > 0.0 ? true : false;
        while (1) {
            min_displacement_until_next_image = velocity[0] != 0.0 ?
                                                separation[0] / velocity[0] + system_length_over_two / fabs(velocity[0])
                                                : INFINITY;
            index_next_image = 0;
            displacement_until_next_image = velocity[1] != 0.0 ?
                                            separation[1] / velocity[1] + system_length_over_two / fabs(velocity[1])
                                            : INFINITY;
            if (displacement_until_next_image < min_displacement_until_next_image) {
                min_displacement_until_next_image = displacement_until_next_image;
                index_next_image = 1;
            }
            displacement_until_next_image = velocity[2] != 0.0 ?
                                            separation[2] / velocity[2] + system_length_over_two / fabs(velocity[2])
                                            : INFINITY;
            if (displacement_until_next_image < min_displacement_until_next_image) {
                min_displacement_until_next_image = displacement_until_next_image;
                index_next_image = 2;
            }
            separation_squared = separation[0] * separation[0] + separation[1] * separation[1]
                                 + separation[2] * separation[2];
            if (separation_dot_velocity_positive) {
                // Norm of separation becomes smaller.
                max_displacement = separation_dot_velocity / velocity_squared;
                if (max_displacement <= min_displacement_until_next_image) {
                    total_displacement += max_displacement;
                    separation[0] -= max_displacement * velocity[0];
                    separation[1] -= max_displacement * velocity[1];
                    separation[2] -= max_displacement * velocity[2];
                    separation_dot_velocity = separation[0] * velocity[0] + separation[1] * velocity[1]
                                              + separation[2] * velocity[2];
                    // Force that separation_dot_velocity is now negative to avoid floating point issues.
                    separation_dot_velocity_positive = false;
                } else {
                    total_displacement += min_displacement_until_next_image;
                    separation[0] -= min_displacement_until_next_image * velocity[0];
                    separation[1] -= min_displacement_until_next_image * velocity[1];
                    separation[2] -= min_displacement_until_next_image * velocity[2];
                    separation[index_next_image] *= -1.0;
                    separation_dot_velocity = separation[0] * velocity[0] + separation[1] * velocity[1]
                                              + separation[2] * velocity[2];
                    separation_dot_velocity_positive = separation_dot_velocity > 0.0 ? true : false;
                }
            } else {
                // Norm of separation becomes bigger.
                max_separation[0] = separation[0] - min_displacement_until_next_image * velocity[0];
                max_separation[1] = separation[1] - min_displacement_until_next_image * velocity[1];
                max_separation[2] = separation[2] - min_displacement_until_next_image * velocity[2];
                current_potential = potential_from_norm(prefactor_product, sqrt(separation_squared));
                max_potential = potential(prefactor_product, max_separation);
                if (potential_change >= max_potential - current_potential) {
                    // Active unit can climb potential hill with remaining potential change.
                    // Reduce potential change and set separation to maximum separation.
                    potential_change -= (max_potential - current_potential);
                    total_displacement += min_displacement_until_next_image;
                    separation[0] = max_separation[0];
                    separation[1] = max_separation[1];
                    separation[2] = max_separation[2];
                    separation[index_next_image] *= -1.0;
                    separation_dot_velocity = separation[0] * velocity[0] + separation[1] * velocity[1]
                                              + separation[2] * velocity[2];
                    separation_dot_velocity_positive = separation_dot_velocity > 0.0 ? true : false;
                } else {
                    new_norm = prefactor_product / (current_potential + potential_change);
                    sqrt_term = separation_dot_velocity * separation_dot_velocity
                                - velocity_squared * (separation_squared - new_norm * new_norm);
                    total_displacement += (separation_dot_velocity + sqrt(sqrt_term)) / velocity_squared;
                    return total_displacement;
                }
            }
        }
    }
}
